### ISSUE TITLE
Feature Request: Adding diff link to recent pabe

### ISSUE THREAD

**Issue by yuki-kimoto** [CPAN-API/metacpan-web#1459]
I want link to diff of before version in recent page.

https://metacpan.org/recent

For example, I want to see the following information easily.

https://metacpan.org/diff/file?target=ZWON/RedisDB-2.40/&source=ZWON/RedisDB-2.39/

recent page only have only distribution link.
https://metacpan.org/release/RedisDB

But I want to see diff information easily in updated module.

Can you add diff link in recent page?

**Comment by oalders** [comment]
@yuki-kimoto we'd certainly accept a patch for this.

**Comment by yuki-kimoto** [comment]
I will try to write this patch in a week.
OK?

**Comment by oalders** [comment]
:+1:

**Comment by yuki-kimoto** [comment]
I add diff link to recent page

https://github.com/yuki-kimoto/metacpan-web/compare/recentdiff

but performance is very slow because I send api by each distribution.
Is there good idea?

**Comment by oalders** [comment]
I think this would be an easier query with a relational database.  Having said that, the api has a /diff endpoint which is basically undocumented.  Given the name of a release it should be able to automatically produce a diff against the previous release:

https://github.com/CPAN-API/cpan-api/blob/master/lib/MetaCPAN/Server/Controller/Diff.pm#L26

For example: https://api.metacpan.org/diff/release/HTML-Restrict

Now that will only work for the very latest authorized releases (unless I'm misreading it), but we could look at tweaking the API endpoint (or adding a new one) in cpan-api to accept an arbitrary release and diff it against the previous release and then have metacpan-web take advantage of this.

That would save the lookups needed to produce the feed as we'd only try to get a diff on demand.

@rwstauner, what do you think?

**Comment by yuki-kimoto** [comment]
If /diff/release is used, it is very easy.

**Comment by ranguard** [comment]
@rwstauner ping

**Comment by rwstauner** [comment]
I think there's room in the api that we could add a `/diff/$author/$release` endpoint
that would diff with the previous release regardless of status.
It would take a few tweaks to existing api code,
and we'd have to add a similar endpoint to the ui, but not too difficult.
This is definitely the way to go for a feature like this (so that you don't have to do any api calls up front).

**Comment by ranguard** [comment]
Thanks @rwstauner 

@yuki-kimoto is this something you'd be interested in working on a patch for ?

**Comment by yuki-kimoto** [comment]
Yes, I want to write code. I think we need API to get latest release diff list.

**Comment by yuki-kimoto** [comment]
I can help you now. What repository do I should see? I think I should access database directory at first to get list of previous releases diff information.

**Comment by rwstauner** [comment]
In [cpan-api](http://github.com/cpan-api/cpan-api):

We'll need a new controller action like this one:
https://github.com/CPAN-API/cpan-api/blob/master/lib/MetaCPAN/Server/Controller/Diff.pm#L26
that takes author and release name as args.

It will need a new "finder" method like these:
https://github.com/CPAN-API/cpan-api/blob/master/lib/MetaCPAN/Document/Release.pm#L290

I think we could use `$c->model('CPAN::Release')->inflate(0)->get({ author => $author, name => $name })` to get the first one, and the new method could use an ES query for "find the first release with a date less than $release->{date}" and pass in the first release.

**Comment by yuki-kimoto** [comment]
At first, I will try capn-api installation today. and will ask some questions.

**Comment by bjakubski** [comment]
ranguard passes following information: "Please see https://github.com/CPAN-API/metacpan-developer for how to get a VM running for developing cpan-api"

**Comment by yuki-kimoto** [comment]
I have an error in installation. My storage rest 4.3GB, but This error seem to indicate disc full.

    $ vagrant up

    There was an error while executing `VBoxManage`, a CLI used by Vagrant
    for controlling VirtualBox. The command and stderr is shown below.

    Command: ["import", "/home/kimoto/.vagrant.d/boxes/mcbase_005/0/virtualbox/box.ovf", "--vsys", "0", "--vmname", "metacpan-developer_default_1409944288947_32066_1429152469594_35730", "--vsys", "0", "--unit", "10", "--disk", "/home/kimoto/VirtualBox VMs/metacpan-developer_default_1409944288947_32066_1429152469594_35730/box-disk1.vmdk"]

    Stderr: 0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
    Interpreting /home/kimoto/.vagrant.d/boxes/mcbase_005/0/virtualbox/box.ovf...
    OK.
    0%...10%...20%...30%...40%...50%...60%...70%...
    Progress state: VBOX_E_FILE_ERROR
    VBoxManage: error: Appliance import failed
    VBoxManage: error: Could not create the imported medium '/home/kimoto/VirtualBox VMs/metacpan-developer_default_1409944288947_32066_1429152469594_35730/box-disk1.vmdk'.
    VBoxManage: error: VMDK: cannot write allocated data block in '/home/kimoto/VirtualBox VMs/metacpan-developer_default_1409944288947_32066_1429152469594_35730/box-disk1.vmdk' (VERR_DISK_FULL)
    VBoxManage: error: Details: code VBOX_E_FILE_ERROR (0x80bb0004), component Appliance, interface IAppliance
    VBoxManage: error: Context: "int handleImportAppliance(HandlerArg*)" at line 779 of file VBoxManageAppliance.cpp

**Comment by rwstauner** [comment]
Strange.

On my machine `~/.vagrant.d/boxes/mcbase_005/0/virtualbox/box-disk1.vmdk` is 1.6G
and the imported disk (`VirtualBox VMs/metacpan-developer_default_*/box-disk1.vmdk`) has grown to 6.6GB.
Mine has extra artifacts downloaded inside, so yours won't be 6.6GB at the start, but it will obviously be at least 1.6GB, so you'll need at least 3.2GB for it all.
After the box is imported the provisioner will clone a few git repos so you'll need even a bit more space than that in the end.
If there are temp files involved in the import that could be an issue.
Maybe some of the disk is reserved for `root`?

Anyway, you can certainly try it again.
You can check `~/.vagrant.d` to make sure there aren't any unnecessary large files left over (there is a tmp dir).

The command `vagrant status` should tell you if the box has been created or not.
It will also show up in the VirtualBox GUI.
If it exists you can try `vagrant destroy` to clean up the old one,
and try another `vagrant up` to try importing again.  It won't re-download the base box,
it just clones the one you already have.

I don't have any other suggestions beyond trying to find more space on your disk.
I haven't heard of people having this issue besides that (though I do hit it myself from time to time).

**Comment by yuki-kimoto** [comment]
thanks.
I try again in other environment which have many space.

**Comment by yuki-kimoto** [comment]
I got some error again in vagrant.

It is hard for me to develop metacpan-api.
I 'm not familiar with vagrant and virtual box. This is very difficult and boring for me.
Sorry, I stop to add recent diff feature to metacpan-api this time.
